### ISSUE TITLE
What Lib i need for Jessie?

### ISSUE THREAD

**Issue by nighteagle1974** [lunixbochs/glshim#146]
After make i became problem to find gl.h after install mesa-common-dev 
After that there is looking for usr/bin/id   -  iGL

Which packge i need for these lib?

Regards

**Comment by lunixbochs** [comment]
Can you post an exact error message?

**Comment by nighteagle1974** [comment]
Yes i will check something now and then i will post it.
I use the RPI2 with an fresh Jessie Image.
First i have to install mesa-common-dev for the missing gl.h.
Now i check after install libgl1-mesa-swx11.... also there is some problems without the 
libgl1-mesa-dri ...

**Comment by nighteagle1974** [comment]
So is not SSH.. i will write Error Message

Bulding 100%
Linking CXX shared library libGLU.so.1
/usr/bin/id : cannot find -iGL
collect2: error: id returned 1 exit status
CMakeFiles/GLU.dir/build.make:2285: recipe for target 'libGLU.so.1' failed

and so on....

**Comment by nighteagle1974** [comment]
Hi,

i have post it.

Regards,

Borris


Am 30.11.2015 18:59, schrieb Ryan Hileman:

**Comment by petevine** [comment]
What's the sillyness with these "i's" instead of "l's" in `/usr/bin/id : cannot find -iGL` ?

**Comment by nighteagle1974** [comment]
So SSH:

[100%] Building CXX object CMakeFiles/GLU.dir/source/libnurbs/internals/flistsorter.cc.o
Linking CXX shared library libGLU.so.1
/usr/bin/ld: cannot find -lGL
collect2: error: ld returned 1 exit status
CMakeFiles/GLU.dir/build.make:2285: recipe for target 'libGLU.so.1' failed
make[2]: *** [libGLU.so.1] Error 1
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/GLU.dir/all' failed
make[1]: *** [CMakeFiles/GLU.dir/all] Error 2
Makefile:76: recipe for target 'all' failed
make: *** [all] Error 2
pi@raspberrypi:~/glues/build $ ^C

**Comment by lunixbochs** [comment]
Did you build and install glshim? You're posting compile errors for glues which is a different project.

**Comment by nighteagle1974** [comment]
???

I have read these README:

https://github.com/lunixbochs/glshim

So what have i to do? I'm confused....

**Comment by lunixbochs** [comment]
You posted a compile error in glues, which is not this project. If you want glshim to be your primary OpenGL provider, try running `make install`.

**Comment by nighteagle1974** [comment]
Hi,
and thanks for the overview... :-)

okay.. i thought i have to use glues also for glshim.
Is there somewhere an manual for glshim... after building them... what have i to do to use the file?
Okay put in lib path and then?

For Background:
My application is for showing the image of an thermografic camera.
It has only 382x288px resolution but without hardware acceleration of the 2D Image i have only 5-6fps in fullscreen. With VC4 i have 30fps. But VC4 have no support for the composite video out. So i want to use the Video out at one time and on other time the hdmi out. Then i have to store the raw-data from the IR-CAM and so i need the cpu-power... there is no gap for rendering image by cpu.
I have build the sources for the IR-CAM and the other way is to reprogram to use OpenGL ES instead of GLX.

Have you an idea?

I can show you the sources from the IR-CAM.

Regards,

**Comment by lunixbochs** [comment]
glshim is a drop-in replacement for desktop libGL. You should be able to just `make install` to use it, or put it in /usr/local/bin, then run your program. To test that glshim is working, try running `glxgears`.

**Comment by nighteagle1974** [comment]
Hi,

so after build and make install successfully i have start glxgears -info and the GL-renderer is Mesa x11.
Okay now i have to disable the installed mesa renderer?

**Comment by lunixbochs** [comment]
Yes. glshim does not depend on Mesa in any way. If you need the Mesa headers, you can copy the glshim/includes/GL folder to /usr/local/include

**Comment by nighteagle1974** [comment]
Hi,

so after remove  mesa i have glxgears -info an Error:
couldn't get an RGB, Double buffered visual.

I have remove mesa-common-dev and libgl1-mesa-dri and libgl1-mesa-swx11.

Any ideas?

**Comment by lunixbochs** [comment]
Was that the only program output? Please paste the entire log.

**Comment by nighteagle1974** [comment]
Hi,

so i begin straight forward an second time.

1. Jessie Image fresh install
2. raspi-config >> Only SSH and Language/Keyboard settings set
3. Network set
4. Mesa-Utils for glxgears install
5. Update and Upgrade Jessie
6. git clone glshim
7. Build glshim
8. make install
9. startx
10. glxgears -info >>

**Comment by lunixbochs** [comment]
No, I mean the program's output.

**Comment by nighteagle1974** [comment]
Yes i know.. but i want an clean fresh install to prevent something mystic..
So after Step 6.1 install cmake i want to build with make GL and becomes following error:

pi@raspberrypi ~/glshim $ make GL
[  0%] Building C object src/CMakeFiles/GL.dir/gl/raster.c.o
In file included from /home/pi/glshim/src/gl/wrap/../types.h:9:0,
                 from /home/pi/glshim/src/gl/wrap/glpack.h:1,
                 from /home/pi/glshim/src/gl/loader.h:10,
                 from /home/pi/glshim/src/gl/raster.c:2:
/home/pi/glshim/src/util/extypes.h:5:22: fatal error: X11/Xlib.h: No such file or directory
 #include <X11/Xlib.h>
                      ^
compilation terminated.
src/CMakeFiles/GL.dir/build.make:54: recipe for target 'src/CMakeFiles/GL.dir/gl/raster.c.o' failed
make[3]: *** [src/CMakeFiles/GL.dir/gl/raster.c.o] Error 1
CMakeFiles/Makefile2:79: recipe for target 'src/CMakeFiles/GL.dir/all' failed
make[2]: *** [src/CMakeFiles/GL.dir/all] Error 2
CMakeFiles/Makefile2:91: recipe for target 'src/CMakeFiles/GL.dir/rule' failed
make[1]: *** [src/CMakeFiles/GL.dir/rule] Error 2
Makefile:151: recipe for target 'GL' failed
make: *** [GL] Error 2
pi@raspberrypi ~/glshim $

So have i install x11 seperatly ???

**Comment by nighteagle1974** [comment]
So glxgears Step 10 of my fresh installation give the same message:
couldn't get an RGB, Double buffered visual.

Where can i have a look for programs output?

**Comment by nighteagle1974** [comment]
I think the problem is dependig from jessie... something with the new systemd ?

**Comment by lunixbochs** [comment]
If the only message from glxgears whatsoever is "couldn't get an RGB, Double buffered visual" then glshim is not being used.

**Comment by nighteagle1974** [comment]
Yes... and what can solve this?
Many people talk about glxgears and this message appear with jessie image.

**Comment by lunixbochs** [comment]
Does es2gears work?

**Comment by nighteagle1974** [comment]
Hi,

which package i have to install?

**Comment by lunixbochs** [comment]
mesa-utils-extra? learn to use apt-file otherwise.

**Comment by nighteagle1974** [comment]
Hi,
i have used apt-cache search...

es2gears bring following:

libEGL warning: DRI2: failed to autheticate
libIGL warning: DRI2: failed to open swrast (path)
EGLUT: failed to initilalize EGL Display

**Comment by lunixbochs** [comment]
You don't have working OpenGL ES drivers, so this isn't a glshim problem.

**Comment by nighteagle1974** [comment]
Ah.. okay.. i thought jessie have the driver onboard. I use keyboard directly on the rpi.

**Comment by nighteagle1974** [comment]
I don't find the OpenGL ES driver...

**Comment by lunixbochs** [comment]
Not a glshim problem :)

**Comment by nighteagle1974** [comment]
I have start it with:
LD_LIBRARY_PATH=/home/pi/glshim/lib LIBGL_FB=1 glxgears -info

and become:

GL_Renderer = VideoCorre IV HW
GL_Version = 1.4 glshim wrapper

**Comment by nighteagle1974** [comment]
I have test it with my application an is running.
X11 is not support OpenGL ES on the RPI!

**Comment by nighteagle1974** [comment]
So for the glshim.. Text ist not supported?
I can see only some signs... but not readable.

**Comment by lunixbochs** [comment]
Text rendered using which method? If you call glXUseXFont, it might have some scaling issues. If you use glBitmap, due to performance reasons all of the text will be rendered on top of the entire scene, so there's no Z order.

**Comment by nighteagle1974** [comment]
So the app use freeglut and in the sources i can see it use glBitmap.
GLUT_BITMAP_8_BY_13 Font is set.

**Comment by nighteagle1974** [comment]
![screen](https://cloud.githubusercontent.com/assets/1252182/11510848/f9f2a038-9865-11e5-83a5-297595592023.jpg)

**Comment by lunixbochs** [comment]
Which glshim branch are you using? unstable or master? Try the other branch.

**Comment by nighteagle1974** [comment]
Which other? Unstable or Master?
i use git clone.. so only one branch is possible?
Example:
git clone https://github.com/lunixbochs/glshim.git

**Comment by nighteagle1974** [comment]
I have the unstable Branch copied. cmake . successfully
So make GL write:
No rule to make target.

??

**Comment by nighteagle1974** [comment]
After test master and rpi branch... they are not working. Message appears after start:

libGL warning, gles is NULL
ERROR: EGL Error detected: EGL_NOT_INITIALIZED (0x3001)
ERROR: EGL Error detected: EGL_NOT_INITIALIZED (0x3001)
Segmentation fault

Only the one branch (actually working) unstable version ist starting

**Comment by petevine** [comment]
You could also try https://github.com/ptitSeb/glshim which can be compiled for Rpi like this:
`cmake . -DBCMHOST=1; make GL`

**Comment by nighteagle1974** [comment]
Hi,

many warnings during make... ?

**Comment by nighteagle1974** [comment]
So i have test out.... https://github.com/ptitSeb/glshim 

I doesn't work.

undefined Symbol:
gl11_client_state_init

Think is from the application an call which is not handled with this version.

**Comment by nighteagle1974** [comment]
The one and oly that start is the unstable branch from lunixbochs.
So good night i have enough for this day.

**Comment by petevine** [comment]
@nighteagle1974 - it [seems](https://github.com/ptitSeb/glshim/commit/399256026bc1f7c20985f74bcd345191067cf715) ptitSeb's heard you, please try again and post your results under the linked commit.

**Comment by lunixbochs** [comment]
@nighteagle1974 

https://github.com/lunixbochs/glshim/blob/unstable/src/gl/raster.c#L49

Try replacing this glRasterPos3f function with this:

    void glRasterPos3f(GLfloat x, GLfloat y, GLfloat z) {
        glWindowPos3f(x, y, z);
    }

Otherwise, can you share your application binary so I can test it myself?

**Comment by nighteagle1974** [comment]
Hi,

i can send you the sources of the app.. but without the IR-CAM you can not start the app.
But you can have a look into the sources to see what going on.
Give me your Mail-Adress then i can send you the sources.

**Comment by lunixbochs** [comment]
lunixbochs@gmail.com

**Comment by nighteagle1974** [comment]
Okay.. delete the mail-adresse here!
I send you the next minutes..

**Comment by nighteagle1974** [comment]
I have replaced glRasterPos3f and build it looks like the same as before..
Nothing is changed...

**Comment by lunixbochs** [comment]
I'll test against your program sometime today and see what's up.

**Comment by nighteagle1974** [comment]
Very Cool... :-)
Where have i to donate?

**Comment by lunixbochs** [comment]
Thanks for the thought! I'm not accepting donations right now, but you could consider supporting the OpenPandora/Pyra project which helped make glshim possible: https://boards.openpandora.org/donate/

**Comment by lunixbochs** [comment]
I reproduced most of your font code in a standalone program based on bitfont.c and I don't actually see the problem here. Can you try running this on your hardware? https://bochs.info/p/nmawz2

**Comment by nighteagle1974** [comment]
Hi,

so i'm not sure what do you mean? 
Please explain.

**Comment by lunixbochs** [comment]
I tested your font rendering code on its own against libglut and glshim, and got this result: http://bochs.info/img/GLUT_bitmap_font_example-20151203-235636.png

I want you to test my version of the code against your hardware to see if it still has the problem.

**Comment by petevine** [comment]
You have to save the test case file to something `.c` e.g. `bitfont.c` ; then compile it:
`gcc bitfont.c -lglut -lGL -lGLU -o bitfont` and finally run the binary.

**Comment by nighteagle1974** [comment]
Hi,

I will  check.

**Comment by nighteagle1974** [comment]
Hi,

so error for compile:

/usr/bin/ld cannot find -lglut
/usr/bin/ld cannot find -lGLU

**Comment by lunixbochs** [comment]
Oh, you can actually remove the `-lGLU` dependency, as I removed that code.

Run ldd on your IR program to see where it's getting libglut from, and add the containing folder as a link path to gcc here using -L/path/to/glut

**Comment by petevine** [comment]
I confirm the broken font display, it's not Rpi specific at all.

**Comment by lunixbochs** [comment]
What are you running it on?

**Comment by nighteagle1974** [comment]
Okay.. what did this mean?

**Comment by petevine** [comment]
Odroid C1, Mali driver.

**Comment by nighteagle1974** [comment]
@petevine 
Is the Ordroid better for performance questions?
I see QuadCore 1,5GHz instead of the RPI 2 with maximum of 1000GHz in stable Overclocking.
So do you use what Linux?
Eventually i can use this also for my IR-CAM.. because after found a solution for rendering i have to implement the saving of the RAW-DATA on USB-Stick.. So i need the CPU Power.. and here any chance for that.

**Comment by petevine** [comment]
Yes, it's much more powerful (you get 1.7Ghz for free and it's around 2x faster in memory performance)

You can choose from the offical Ubuntu or unofficial Debian or Arch images.

**Comment by nighteagle1974** [comment]
Good advantages.. And what the disadvantages? :-)

**Comment by lunixbochs** [comment]
@petevine can you post a screenshot of my example app on the Odroid?

**Comment by petevine** [comment]
@lunixbochs - see attachment.
![bitfont](https://cloud.githubusercontent.com/assets/12054457/11601357/f1a807c8-9ad2-11e5-8dc2-de3f1ef42cea.jpg)

@nighteagle1974 

The disadvantages (still extant) are mainly these: crappy Mali driver performance and USB issues when mixing 2.0 and 1.1 devices on the same host (there's a second OTG host so it's possible to work around this issue). 

The usb issues are currently being worked on via FIQ patches so very soon C1 is going to be almost problem-free.

**Comment by nighteagle1974** [comment]
@lunixbochs bitfont.c for me short bit.c

pi@raspberrypi ~/bit_c $ ls
bit.c
pi@raspberrypi ~/bit_c $ gcc bit.c -lglut -lGL -lGLU -o bit
/usr/lib/gcc/arm-linux-gnueabihf/4.9/../../../arm-linux-gnueabihf/crt1.o: In function `_start':
/build/glibc-IRmfEu/glibc-2.19/csu/../ports/sysdeps/arm/start.S:119: undefined reference to `main'
collect2: error: ld returned 1 exit status
pi@raspberrypi ~/bit_c $ gcc bit.c -lglut -lGL -o bit
/usr/lib/gcc/arm-linux-gnueabihf/4.9/../../../arm-linux-gnueabihf/crt1.o: In function `_start':
/build/glibc-IRmfEu/glibc-2.19/csu/../ports/sysdeps/arm/start.S:119: undefined reference to `main'
collect2: error: ld returned 1 exit status
pi@raspberrypi ~/bit_c $

**Comment by nighteagle1974** [comment]
I'm blinding and confused.. i have a break....

**Comment by lunixbochs** [comment]
@petevine thanks! can you replace the output function with this (to remove the background rendering) and post a new screenshot?

    void output(int x, int y, char *string) {
        unsigned char *text = (unsigned char *)string;
        unsigned char fg[4] = {255, 255, 255, 255};

        glColor4ub(fg[0], fg[1], fg[2], fg[3]);
        glWindowPos2i(x, y);
        glutBitmapString(font, text);
    }

**Comment by petevine** [comment]
Done!
![fixed](https://cloud.githubusercontent.com/assets/12054457/11602175/8c53253c-9ad8-11e5-87f8-382a6fa8c8a6.jpg)

**Comment by lunixbochs** [comment]
@nighteagle1974 is the text positioning correct in your image, just the display is wrong?

**Comment by petevine** [comment]
Oops, sorry! I forgot to change my LD_LIBRARY_PATH. That was ptitSeb's glshim doing the work.
It didn't make a difference using yours.

**Comment by lunixbochs** [comment]
I think this is a couple of things. Pixel transparency, origin vertical flip, and upside down bitmap rendering. I'll play with it and probably patch tonight.

**Comment by lunixbochs** [comment]
Can someone try the bitmap-test branch against either program and post a screenshot?

**Comment by petevine** [comment]
No change really. For some reason the example binary doesn't work @32bpp (under any glshim)
![bitmap](https://cloud.githubusercontent.com/assets/12054457/11606202/cb5707c2-9b18-11e5-8494-46f8219fcb64.jpg)

**Comment by nighteagle1974** [comment]
@lunixbochs 
No if you can see here on my picture Posts:
https://boards.openpandora.org/topic/11506-glshim/?page=24

Correct : The Text on the top right corner.. so there is an Textfield rectangle.

**Comment by nighteagle1974** [comment]
Hi,
today i have test something and all test show the scroling text.

**Comment by petevine** [comment]
You mean in lunixbochs's glshim or rather ptitSeb's again?
